### PR TITLE
tsweb: accept a function to call before request handling

### PR DESCRIPTION
**Pull request by knyar**
To complement the existing `onCompletion` callback, which is called after request handler.

Updates tailscale/corp#17075

### CODE REVIEW

File: tsweb/tsweb.go [hunk 1]
```diff
@@ -250,6 +250,9 @@ type HandlerOptions struct {
 	// for each bucket based on the contained parameters.
 	BucketedStats *BucketedStatsOptions
 
+	// OnStart is called before ServeHTTP is called.
```
Comment (bradfitz) [inline tsweb/tsweb.go:253]: document as optional

File: tsweb/tsweb_test.go [hunk 2]
```diff
@@ -485,8 +485,16 @@ func TestStdHandler(t *testing.T) {
...
+			onStartRecord := &AccessLogRecord{}
+			onCompletionRecord := &AccessLogRecord{}
 			rec := noopHijacker{httptest.NewRecorder(), false}
-			h := StdHandler(test.rh, HandlerOptions{Logf: logf, Now: clock.Now, OnError: test.errHandler})
+			h := StdHandler(test.rh, HandlerOptions{
+				Logf:         logf,
+				Now:          clock.Now,
+				OnError:      test.errHandler,
+				OnStart:      func(r *http.Request, alr AccessLogRecord) { *onStartRecord = alr },
+				OnCompletion: func(r *http.Request, alr AccessLogRecord) { *onCompletionRecord = alr },
```
Comment (bradfitz) [inline tsweb/tsweb_test.go:496]: are these pointers or values? why the mix?

Comment (knyar) [reply tsweb/tsweb_test.go:496]: Not sure I follow - existing `OnCompletion` passed `AccessLogRecord` by value, so I've added `OnStart` to do the same. It's not a big struct, and `tsweb` continues to mutate it after calling these hooks, so I think passing by value makes sense, but I am happy to change it.

Comment (bradfitz) [reply tsweb/tsweb_test.go:496]: why are they pointers on line 488 and 489? why not just values there?

Comment (knyar) [reply tsweb/tsweb_test.go]: Oh right, these were left from a slightly different version of the test. Changed to values.

File: tsweb/tsweb_test.go [hunk 3]
```diff
@@ -502,6 +509,12 @@ func TestStdHandler(t *testing.T) {
 				}
 				return e.Error()
 			})
+			if onStartRecord.Time.IsZero() {
+				t.Error("onStart callback not called")
+			}
```
Comment (icio) [inline tsweb/tsweb_test.go:514]: You could compare onStartRecord to test.wantLog but specify the fields only known at completion with cmpopts.IgnoreFields().

**Review by icio — APPROVED** [review]
LGTM

File: tsweb/tsweb.go [hunk 4]
```diff
@@ -250,19 +250,25 @@ type HandlerOptions struct {
 	// for each bucket based on the contained parameters.
 	BucketedStats *BucketedStatsOptions
 
+	// OnStart is called before ServeHTTP is called. Optional.
```
Comment (icio) [inline tsweb/tsweb.go:253]:
```suggestion
	// OnStart is called inline before ServeHTTP is called. Optional.
```

File: tsweb/tsweb.go [hunk 5]
```diff
@@ -250,19 +250,25 @@ type HandlerOptions struct {
...
+	OnStart OnStartFunc
+
 	// OnError is called if the handler returned a HTTPError. This
 	// is intended to be used to present pretty error pages if
 	// the user agent is determined to be a browser.
 	OnError ErrorHandlerFunc
 
 	// OnCompletion is called when ServeHTTP is finished and gets
```
Comment (icio) [inline tsweb/tsweb.go:261]:
```suggestion
	// OnCompletion is called inline when ServeHTTP is finished and gets
```

**Review by icio — COMMENTED** [review]
Maybe worth noting that OnStart/OnCompletion are inline, blocking callbacks.

File: tsweb/tsweb_test.go (hunk 3 above)
Comment (knyar) [reply tsweb/tsweb_test.go]: Done

**Pull request merged**